text/0131-pass-native-sdk-spans-to-hybrid.md

```diff
@@ -0,0 +1,174 @@
+in a data structure ordered by span timestamp and use a binary search to find the last index of the
+range of spans to delete, which makes inserting a span `O(log n)`. Furthermore, the native SDKs
+offer a callback when all ongoing spans are finished to support `waitForChildren`.
+
+### Pros <a name="option-1-pros"></a>
+
+1. Less cross hybrid layer communication than [option 3](#option-3).
+
+### Cons <a name="option-1-cons"></a>
+
+1. With single-span ingestion, we don't want to have a self-clearing cache. Instead, every span
+should end up in Sentry.
+2. Adding spans to the cache is `O(log n)`.
+
+## Option 2: Spans Aggregator <a name="option-2"></a>
+
+The native SDKs use the [SpansAggregator](/text/0126-sdk-spans-aggregator.md) to cache the spans.
+The SpansAggregator doesn't pass the spans down to the transport for sending. Instead, when its
+timeout exceeds or aggregated spans exceed the maximum byte size, the SpansAggregator passes the
+spans up to the hybrid layer. The SpansAggregator implements a new method for retrieving all spans
+so that hybrid SDKs can attach spans to transactions bound to the scope or active spans.
+Furthermore, until single-span ingestion lands, the native SDKs offer a callback when all ongoing
+spans are finished to support `waitForChildren` and provide hooks to start and stop the
+SpansAggregator, same as [option 1](#option-1).
+
+The following specification is written in the [Gherkin syntax](https://cucumber.io/docs/gherkin/reference/)
+to explain how the solution reacts to multiple edge cases. To simplify the requirements, the
+specification only deals with active spans, which are similar to transactions bound to the scope. A
+few points to consider for the specification:
+
+1. An active span always refers to the hybrid SDKs because native SDKs don't have active spans when
+in hybrid mode.
+2. When the hybrid SDKs add the spans from the native layer to their SpansAggregator, the spans must
+pass through all hooks as when finishing a hybrid span.
+3. The hybrid SpansAggregator keeps its behavior. No changes are required.
+4. After the hybrid SDKs pull spans from the native SpansAggregator, the native SpansAggregator
+removes all its spans.
+
+```Gherkin
+Scenario: No active span, native SpansAggregator flushes
+    Given no active span
+    When the native SpansAggregator has to flush all its spans
+    Then native SDK pushes all native spans up to the hybrid SDK
+    And doesn't send its native spans directly to Sentry
+    And the hybrid SDK adds all native spans to its SpansAggregator
+
+Scenario: Active span, native SpansAggregator flushes
+    Given an active span
+    When the native SpansAggregator has to flush all its spans
+    Then native SDK pushes all native spans up to the hybrid SDK
+    And the hybrid SDK filters for spans belonging to the active span 
+    And adds these spans to the active span
+    And adds the non matching spans to the hybrid SpansAggregator
```

Comment:
Does option 2 counts with single-span ingestion, if we consider the active span being a transaction, what would happen with the non-matched spans?

Reply:
In that case, we would throw them away as we do right now.